module/VuFind/src/VuFind/Db/Service/RecordService.php

```diff
@@ -0,0 +1,179 @@
+        $records = $query->getResult();
+        return count($records) > 0 ? current($records) : false;
+    }
+
+    /**
+     * Find records by ids
+     *
+     * @param array  $ids    Record IDs
+     * @param string $source Record source
+     *
+     * @return array Array of record objects found
+     */
+    public function findRecords($ids, $source)
+    {
+        if (empty($ids)) {
+            return [];
+        }
+
+        $dql = "SELECT r "
+            . "FROM " . $this->getEntityClass(Record::class) . " r "
+            . "WHERE r.recordId IN (:ids) AND r.source = :source";
+        $parameters = compact('ids', 'source');
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $records = $query->getResult();
+        return $records;
+    }
+
+    /**
+     * Update an existing entry in the record table or create a new one
+     *
+     * @param string $id      Record ID
+     * @param string $source  Data source
+     * @param string $rawData Raw data from source
+     *
+     * @return Record|bool
+     */
+    public function updateRecord($id, $source, $rawData)
+    {
+        $record = $this->findRecord($id, $source);
+        if (!$record) {
+            $record = $this->createEntity();
+        }
+        $record->setRecordId($id)
+            ->setSource($source)
+            ->setData(serialize($rawData))
+            ->setVersion(\VuFind\Config\Version::getBuildVersion())
+            ->setUpdated(new \DateTime());
+        try {
+            $this->persistEntity($record);
+        } catch (\Exception $e) {
+            $this->logError('Could not save record: ' . $e->getMessage());
+            return false;
+        }
+        return $record;
+    }
+
+    /**
+     * Clean up orphaned entries (i.e. entries that are not in favorites anymore)
+     *
+     * @return int Number of records deleted
+     */
+    public function cleanup()
+    {
+        $dql = "SELECT r.id "
+            . "FROM " . $this->getEntityClass(Record::class) . " r "
+            . "JOIN " . $this->getEntityClass(Resource::class) . " re "
+            . "WITH r.recordId = re.recordId AND r.source = re.source "
+            . "LEFT JOIN " . $this->getEntityClass(UserResource::class) . " ur "
+            . "WITH re.id = ur.resource "
+            . "WHERE ur.id IS NULL";
+        $query = $this->entityManager->createQuery($dql);
+        $ids = $query->getResult();
+        $dql = "DELETE FROM " . $this->getEntityClass(Record::class) . " r "
+            . "WHERE r.id IN (:ids)";
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters(compact('ids'));
+        $query->execute();
+        return count($ids);
+    }
+
+    /**
+     * Delete a record by source and id
+     *
+     * @param string $source Record source
+     * @param string $id     Record ID
+     *
+     * @return mixed
+     */
+    public function deleteRecord($source, $id)
```

Comment:
It might make sense to reverse the arguments of this function to `$id, $source` for consistency with other methods in the class which take these two parameters.